utility/session_test.go

```diff
@@ -0,0 +1,531 @@
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/messaging"
+	"github.com/pokt-network/pocket/utility/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gonum.org/v1/gonum/stat/combin"
+)
+
+// TECHDEBT: Geozones are not current implemented, used or tested
+
+func TestSession_GetSession_SingleFishermanSingleServicerBaseCase(t *testing.T) {
+	// Test parameters
+	height := int64(1)
+	relayChain := test_artifacts.DefaultChains[0]
+	geoZone := "unused_geo"
+	numFishermen := 1
+	numServicers := 1
+	expectedSessionId := "5acf559f1a3faf3bea7eb692fe51bc1e2e5fb687ede0a6daa7d42399da4aa82b" // needs to be manually updated if business logic changes
+
+	runtimeCfg, utilityMod, _ := prepareEnvironment(t, 5, numServicers, 1, numFishermen)
+
+	// Sanity check genesis
+	require.Len(t, runtimeCfg.GetGenesis().Applications, 1)
+	app := runtimeCfg.GetGenesis().Applications[0]
+	require.Len(t, runtimeCfg.GetGenesis().Fishermen, 1)
+	fish := runtimeCfg.GetGenesis().Fishermen[0]
+	require.Len(t, runtimeCfg.GetGenesis().Servicers, 1)
+	servicer := runtimeCfg.GetGenesis().Servicers[0]
+
+	// Verify some of the session defaults
+	session, err := utilityMod.GetSession(app.Address, height, relayChain, geoZone)
+	require.NoError(t, err)
+	require.Equal(t, expectedSessionId, session.Id)
+	require.Equal(t, height, session.SessionHeight)
+	require.Equal(t, int64(1), session.SessionNumber)
+	require.Equal(t, int64(1), session.NumSessionBlocks)
+	require.Equal(t, relayChain, session.RelayChain)
+	require.Equal(t, geoZone, session.GeoZone)
+	require.Equal(t, app.Address, session.Application.Address)
+	require.Len(t, session.Servicers, numServicers)
+	require.Equal(t, servicer.Address, session.Servicers[0].Address)
+	require.Len(t, session.Fishermen, numFishermen)
+	require.Equal(t, fish.Address, session.Fishermen[0].Address)
+}
+
+func TestSession_GetSession_ApplicationInvalid(t *testing.T) {
+	runtimeCfg, utilityMod, _ := prepareEnvironment(t, 5, 1, 1, 1)
+
+	// Verify there's only 1 app
+	require.Len(t, runtimeCfg.GetGenesis().Applications, 1)
+	app := runtimeCfg.GetGenesis().Applications[0]
+
+	// Create a new app address
+	pk, err := crypto.GeneratePrivateKey()
+	require.NoError(t, err)
+
+	// Verify that the one app in the genesis is not the one we just generated
+	addr := pk.Address().String()
+	require.NotEqual(t, app.Address, addr)
+
+	// Expect no error trying to get a session for the real application
+	_, err = utilityMod.GetSession(app.Address, 1, test_artifacts.DefaultChains[0], "unused_geo")
+	require.NoError(t, err)
+
+	// Expect an error trying to get a session for an unstaked chain
+	_, err = utilityMod.GetSession(addr, 1, "chain", "unused_geo")
+	require.Error(t, err)
+
+	// Expect an error trying to get a session for a non-existent application
+	_, err = utilityMod.GetSession(addr, 1, test_artifacts.DefaultChains[0], "unused_geo")
+	require.Error(t, err)
```

Comment:
It looks like there are several kinds of errors which can be returned. Could you imagine a scenario where these assertions potentially pass for the wrong reason? Wdyt about either adding `TECHDEBT` comments or refactoring the error messages so they can be tested for by identity (until we potentially decide otherwise; #519)?

Reply:
bump

Reply:
Added `// TECHDEBT(#519): Add custom error types depending on the type of issue that occurred and assert on them in the unit tests.`